cpp/src/strings/combine/concatenate.cu

```diff
@@ -0,0 +1,294 @@
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto const num_columns = strings_columns.num_columns();
+  CUDF_EXPECTS(num_columns > 0, "At least one column must be specified");
+  // check all columns are of type string
+  CUDF_EXPECTS(std::all_of(strings_columns.begin(),
+                           strings_columns.end(),
+                           [](auto c) { return c.type().id() == type_id::STRING; }),
+               "All columns must be of type string");
+  if (num_columns == 1)  // single strings column returns a copy
+    return std::make_unique<column>(*(strings_columns.begin()), stream, mr);
+  auto const strings_count = strings_columns.num_rows();
+  if (strings_count == 0)  // empty begets empty
+    return detail::make_empty_strings_column(stream, mr);
+
+  CUDF_EXPECTS(separator.is_valid(), "Parameter separator must be a valid string_scalar");
+  string_view d_separator(separator.data(), separator.size());
+  auto d_narep = get_scalar_device_view(const_cast<string_scalar&>(narep));
+
+  // Create device views from the strings columns.
+  auto d_table = table_device_view::create(strings_columns, stream);
+  concat_strings_fn fn{*d_table, d_separator, d_narep};
+  auto children = make_strings_children(fn, strings_count, 0, stream, mr);
+
+  // create resulting null mask
+  auto valid_mask = cudf::detail::valid_if(
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(strings_count),
+    [d_table = *d_table, d_narep] __device__(size_type idx) {
+      bool null_element = thrust::any_of(
+        thrust::seq, d_table.begin(), d_table.end(), [idx](auto col) { return col.is_null(idx); });
+      return (!null_element || d_narep.is_valid());
+    },
+    stream,
+    mr);
+  auto& null_mask       = valid_mask.first;
+  auto const null_count = valid_mask.second;
+
+  return make_strings_column(strings_count,
+                             std::move(children.first),
+                             std::move(children.second),
+                             null_count,
+                             std::move(null_mask),
+                             stream,
+                             mr);
+}
+
+namespace {
+
+/**
+ * @brief Concatenate strings functor using multiple separators.
+ *
+ * A unique separator is provided for each row along with a string to use
+ * when a separator row is null `d_separator_narep`. The `d_narep` is
+ * used in place of a null entry in the strings columns.
+ */
+struct multi_separator_concat_fn {
+  table_device_view const d_table;
+  column_device_view const d_separators;
+  string_scalar_device_view const d_separator_narep;
+  string_scalar_device_view const d_narep;
+  offset_type* d_offsets{};
+  char* d_chars{};
+
+  __device__ void operator()(size_type idx)
+  {
+    bool const all_nulls =
+      thrust::all_of(thrust::seq, d_table.begin(), d_table.end(), [idx](auto const& col) {
```

Comment:
For `thrust::seq` `count_if` can be a little less efficient since all elements must be tested whereas `all_of` can return early once a non-null element is encountered. I did a quick measurement and found no significant performance difference. I think `all_of` more clearly conveys the intent here.

Reply:
Ah, I see, probably because the execution here is `thrust::seq` instead of parallel.